tokio/src/time/driver/mod.rs

```diff
@@ -91,6 +91,13 @@ pub(crate) struct Driver<P: Park + 'static> {
 
     /// Parker to delegate to
     park: P,
+
+    // When test-util is enabled, this tracks if the driver was woken before
+    // entering park. While it may look racy, it only has any effect when the
+    // clock is paused and pausing the clock is restricted to a single-threaded
+    // runtime.
+    #[cfg(feature = "test-util")]
+    did_wake: Arc<AtomicBool>,
```

Comment:
I don't really understand the comment on this. When is the time driver woken?

Reply:
Any time Unpark::unpark() is called.

Reply:
So is that when a timer elapses?

Reply:
Not exactly. When a timer elapses, the associated Waker is notified.

Wakers for the tokio runtime are implemented by scheduling the associated task then calling `unpark()` on the runtime. Calling unpark will interrupt an epoll_wait call within the runtime (wakeup the io::driver).

In the time driver, I am intercepting this call as a signal to tell the driver that "something" woke up the runtime, i.e. the timeout passed to `park()` did not elapse.

Futures passed to `Runtime::block_on` are not scheduled on the runtime the same way spawned futures are. When the basic_scheduler polls all scheduled tasks, it calls park(None), the timer driver changes None -> the duration until the next timeout elapses. If the task passed to `block_on` is awaiting on `task::yield_now()`, it will first have called `cx.waker().wake_by_ref()` which, in turn, will call `unpark()`. This is how the basic_scheduler avoids blocking forever when it calls `park(None)`.

Reply:
That makes sense. In particular, the important piece is that `did_wake` tracks whether the timeout elapsed, or something else happened.

Can you add this to the comment in the PR?